content/post/2022-04-25-PHP-database-result-iteration.md

```diff
@@ -0,0 +1,164 @@
+---
+title: Best practices in PHP when iterating over database results
+date: 2022-04-25
+tags:
+  - PHP
+  - database
+  - history
+  - patterns
+description: "How the default method for iterating over database results
+is stuck in 2005, goes against best practices and what to do about it."
+---
+
+Many PHP database libraries use a `while` loop with assignment to iterate
+over database results. In this article I'll show why this pattern is so
+pervasive, why it opens the door for errors and what
+you can use instead.
+
+<!--more-->
+
+This is a typical piece of PHP code, using [PDO](https://www.php.net/manual/en/book.pdo.php) for accessing the database and iterating over the result:
+
+```php
+$result = $db->query("SELECT id, name from my_table");
+while ($row = $result->fetch()) {
```

Comment:
There is a really good (historical) reason why people wanted code like this. The idea is that this loop will only hold 1 database row in memory at a time, but never the entire result set, which might be millions of rows. This is how the code reads, isn't it? `query()` returns only a pointer or resource identifier or whatever, but that's it. `fetch()` transfers 1 row from the MySQL server to the PHP process. And the next row. And so on.

This is why you will find code like this.
```php
$res = $db->query( 'SELECT * FROM table ORDER BY time DESC' );
if ( $row = $res->fetch() ) {
    // Process the first, most recent result, but nothing else
}
```
People assume this is as optimized as it can be.

Unfortunately this is not what happens. (At least not when I last checked years ago.) `query()` reads everything into memory.

This kind of makes sense when you think about it:
* How should `numRows()` know what to say without reading everything first?
* The MySQL server would need to hold a copy of the table as it was when the query was executed to make sure all following `fetch()` calls return something that forms a consistent result set. I guess the idea was "when we need to hold a copy anyway we can as well do this in the PHP extension".

I avoid this pattern since then because of this broken promise it seems to give, but doesn't – probably can't – give.